packages/tldraw/src/lib/shapes/arrow/arrowLabel.ts

```diff
@@ -309,6 +306,47 @@ export function getArrowLabelPosition(editor: Editor, shape: TLArrowShape) {
 	return { box: Box.FromCenter(labelCenter, labelSize), debugGeom }
 }
 
+function getClampedPosition(
+	editor: Editor,
+	shape: TLArrowShape,
+	range: { start: number; end: number },
+	arrowheadInfo: ArrowheadInfo
+) {
+	const { hasEndArrowhead, hasEndBinding, hasStartBinding, hasStartArrowhead } = arrowheadInfo
+	const arrowLength = getArrowLength(editor, shape)
+	let clampedPosition = clamp(
+		shape.props.labelPosition,
+		hasStartArrowhead || hasStartBinding ? range.start : 0,
+		hasEndArrowhead || hasEndBinding ? range.end : 1
+	)
+	const minLength = 100
+	const maxLength = 1500
+	const minSnapDistance = 0.03
+	const maxSnapDistance = 0.005
+
+	let snapDistance = minSnapDistance
+	if (arrowLength > minLength) {
+		if (arrowLength < maxLength) {
+			snapDistance = lerp(
+				minSnapDistance,
+				maxSnapDistance,
+				(arrowLength - minLength) / (maxLength - minLength)
+			)
+		} else {
+			snapDistance = maxSnapDistance
+		}
+	}
+
+	const wasClamped = clampedPosition >= 0.5 - snapDistance && clampedPosition <= 0.5 + snapDistance
+
+	if (wasClamped) {
+		clampedPosition = 0.5
+		editor.updateShape({ ...shape, props: { ...shape.props, labelPosition: clampedPosition } })
```

Comment:
I think the big issue is that we shouldn't update the shape (that's why things will drift when changing the endpoints). `getClampedPosition` shouldn't have a side-effect like this basically. (if you `get` something it shouldn't _change_ the underlying thing)